# Segmenting and Clustering Neighborhoods in Toronto

#### Generate Dataframe

In [173]:
import pandas as pd
from bs4 import BeautifulSoup
import requests

wiki = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
wikipedia_page = requests.get(wiki)

df_raw = pd.read_html(wikipedia_page.content, header=0)[0]

df_raw.head()

Postal Code           Borough               Neighborhood
0         M1A      Not assigned                        NaN
1         M2A      Not assigned                        NaN
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

#### Remove Borough that is not assigned 

In [174]:
# Ignore cells with a borough that is Not assigned.

df_new = df_raw[df_raw.Borough != 'Not assigned']

df_new.head()

Postal Code           Borough                                 Neighborhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

#### Neighborhood that is not assigned to take on name of Borough

In [175]:
# If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough.


df_new.loc[df_new.Neighborhood == 'Not assigned']
df_new.Neighborhood.replace('Not assigned',df_new.Borough,inplace=True)
df_new.head()

Postal Code           Borough                                 Neighborhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

#### Combining Neighborhood with the same Postal Code

In [176]:
# Combining neighbourhood with the same Postal Code


df_toronto = df_new.groupby(['Postal Code', 'Borough'])['Neighborhood'].apply(lambda x: ', '.join(x))
df_toronto = df_toronto.reset_index()
df_toronto.rename(columns = {'Neighbourhood':'Neighborhood'}, inplace = True)
df_toronto.head()

Postal Code      Borough                            Neighborhood
0         M1B  Scarborough                          Malvern, Rouge
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek
2         M1E  Scarborough       Guildwood, Morningside, West Hill
3         M1G  Scarborough                                  Woburn
4         M1H  Scarborough                               Cedarbrae

#### Shape of the df

In [177]:
df_toronto.shape

(103, 3)

## Assumptions made:<br/>
1) Boroughs that are not assigned are removed for analysis<br/>
2) Neighborhoods that are not assigned takes no the name of the borough<br/>
3) Neighbourhoods with the same postal code are combined<br/>
<br/>
**There are 103 unique Postal Code in Toronto**

## Get the latitude and the longitude coordinates of each neighborhood

use csv method being the more reliable method

In [178]:
url = 'http://cocl.us/Geospatial_data'
df_geo=pd.read_csv(url)
df_geo.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [179]:
#check the shape ofthe csv file
df_geo.shape

(103, 3)

In [180]:
df_toronto.head()

Postal Code      Borough                            Neighborhood
0         M1B  Scarborough                          Malvern, Rouge
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek
2         M1E  Scarborough       Guildwood, Morningside, West Hill
3         M1G  Scarborough                                  Woburn
4         M1H  Scarborough                               Cedarbrae

#### Merge the Latitude Data and Longtitude Data to the neighborhoods

In [184]:
df_toronto_merge=df_toronto.merge(df_geo, on='Postal Code')
df_toronto_merge.head()

Postal Code      Borough                            Neighborhood   Latitude  \
0         M1B  Scarborough                          Malvern, Rouge  43.806686   
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2         M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3         M1G  Scarborough                                  Woburn  43.770992   
4         M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

## Use the foursquere API to segment and cluster the neighborhoods of Toronto

In [185]:
!conda install -c conda-forge geocoder --yes
import geocoder
from geopy.geocoders import Nominatim 

address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - geocoder


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    certifi-2020.4.5.1         |   py36h9f0ad1d_0         151 KB  conda-forge
    geocoder-1.38.1            |             py_1          53 KB  conda-forge
    ratelim-0.1.6              |             py_2           6 KB  conda-forge
    ca-certificates-2020.4.5.1 |       hecc5488_0         146 KB  conda-forge
    python_abi-3.6             |          1_cp36m           4 KB  conda-forge
    openssl-1.1.1g             |       h516909a_0         2.1 MB  conda-forge
    ------------------------------------------------------------
                                           Total:         2.5 MB

The following NEW packages will be INSTALLED:

    geocoder:        1.38.1-py_1       conda-forge
    python_abi:    

In [189]:
!conda install -c conda-forge folium



Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - folium


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    branca-0.4.1               |             py_0          26 KB  conda-forge
    folium-0.11.0              |             py_0          61 KB  conda-forge
    ------------------------------------------------------------
                                           Total:          87 KB

The following NEW packages will be INSTALLED:

    branca: 0.4.1-py_0  conda-forge
    folium: 0.11.0-py_0 conda-forge


branca-0.4.1         | 26 KB     | ##################################### | 100% 
folium-0.11.0        | 61 KB     | ##################################### | 100% 
Preparing transaction: done
Verifying transaction: done
Executing transaction: done


NameError: name 'folium' is not defined

In [193]:
import folium

# create map of Toronto using latitude and longitude values
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_toronto_merge['Latitude'], df_toronto_merge['Longitude'], df_toronto_merge['Borough'], df_toronto_merge['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        ).add_to(map_Toronto)  
    
map_Toronto

#### Define Foursquare Credentials and Version

In [194]:
CLIENT_ID = 'VFJBMTZRUJWJUETLOF1ZOKT54RGUTCTOCZSR3JCBRSCG1JEK' # your Foursquare ID
CLIENT_SECRET = '4TVUF2ITDPWVCM0THUKPBSETS5SORFHMFTXPOCJQOK0X454R' # your Foursquare Secret
VERSION = '20180604' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: VFJBMTZRUJWJUETLOF1ZOKT54RGUTCTOCZSR3JCBRSCG1JEK
CLIENT_SECRET:4TVUF2ITDPWVCM0THUKPBSETS5SORFHMFTXPOCJQOK0X454R


In [195]:
#df_toronto.loc[0, 'Neighbourhood']

In [197]:
neighborhood_latitude = df_toronto_merge.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = df_toronto_merge.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = df_toronto.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Malvern, Rouge are 43.806686299999996, -79.19435340000001.


#### Create the GET request URL 

In [198]:
LIMIT = 100
radius = 500
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url

'https://api.foursquare.com/v2/venues/explore?&client_id=VFJBMTZRUJWJUETLOF1ZOKT54RGUTCTOCZSR3JCBRSCG1JEK&client_secret=4TVUF2ITDPWVCM0THUKPBSETS5SORFHMFTXPOCJQOK0X454R&v=20180604&ll=43.806686299999996,-79.19435340000001&radius=500&limit=100'

In [199]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5ed22039dd0f85001b0bb1db'},
 'response': {'warning': {'text': "There aren't a lot of results near you. Try something more general, reset your filters, or expand the search area."},
  'headerLocation': 'Malvern',
  'headerFullLocation': 'Malvern, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 1,
  'suggestedBounds': {'ne': {'lat': 43.8111863045, 'lng': -79.18812958073042},
   'sw': {'lat': 43.80218629549999, 'lng': -79.2005772192696}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4bb6b9446edc76b0d771311c',
       'name': 'Wendy’s',
       'location': {'crossStreet': 'Morningside & Sheppard',
        'lat': 43.80744841934756,
        'lng': -79.19905558052072,
        'labeledLatLngs': [{'label': 'display'

In [200]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

 #### only 1 response for our first entry

In [201]:
import json
from pandas.io.json import json_normalize

venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name            categories        lat        lng
0  Wendy’s  Fast Food Restaurant  43.807448 -79.199056

#### Obtaining the venues from all neighbourhoods in Toronto

In [203]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [205]:
toronto_venues = getNearbyVenues(names=df_toronto_merge['Neighborhood'],
                                   latitudes=df_toronto_merge['Latitude'],
                                   longitudes=df_toronto_merge['Longitude']
                                  )

Malvern, Rouge
Rouge Hill, Port Union, Highland Creek
Guildwood, Morningside, West Hill
Woburn
Cedarbrae
Scarborough Village
Kennedy Park, Ionview, East Birchmount Park
Golden Mile, Clairlea, Oakridge
Cliffside, Cliffcrest, Scarborough Village West
Birch Cliff, Cliffside West
Dorset Park, Wexford Heights, Scarborough Town Centre
Wexford, Maryvale
Agincourt
Clarks Corners, Tam O'Shanter, Sullivan
Milliken, Agincourt North, Steeles East, L'Amoreaux East
Steeles West, L'Amoreaux West
Upper Rouge
Hillcrest Village
Fairview, Henry Farm, Oriole
Bayview Village
York Mills, Silver Hills
Willowdale, Newtonbrook
Willowdale, Willowdale East
York Mills West
Willowdale, Willowdale West
Parkwoods
Don Mills
Don Mills
Bathurst Manor, Wilson Heights, Downsview North
Northwood Park, York University
Downsview
Downsview
Downsview
Downsview
Victoria Village
Parkview Hill, Woodbine Gardens
Woodbine Heights
The Beaches
Leaside
Thorncliffe Park
East Toronto, Broadview North (Old East York)
The Danforth West, 

In [206]:
print(toronto_venues.shape)
toronto_venues.head()

(2136, 7)


Neighbourhood  Neighborhood Latitude  \
0                          Malvern, Rouge              43.806686   
1  Rouge Hill, Port Union, Highland Creek              43.784535   
2  Rouge Hill, Port Union, Highland Creek              43.784535   
3       Guildwood, Morningside, West Hill              43.763573   
4       Guildwood, Morningside, West Hill              43.763573   

   Neighborhood Longitude                           Venue  Venue Latitude  \
0              -79.194353                         Wendy’s       43.807448   
1              -79.160497           Royal Canadian Legion       43.782533   
2              -79.160497  Scarborough Historical Society       43.788755   
3              -79.188711                  RBC Royal Bank       43.766790   
4              -79.188711               G & G Electronics       43.765309   

   Venue Longitude        Venue Category  
0       -79.199056  Fast Food Restaurant  
1       -79.163085                   Bar  
2       -79.162438        History Museum  
3       -79.191151                  Bank  
4       -79.191537     Electronics Store

#### Number of venues for each Neighbourhood

In [207]:
toronto_venues.groupby('Neighbourhood').count()

Neighborhood Latitude  \
Neighbourhood                                                               
Agincourt                                                               4   
Alderwood, Long Branch                                                 11   
Bathurst Manor, Wilson Heights, Downsview North                        21   
Bayview Village                                                         4   
Bedford Park, Lawrence Manor East                                      23   
Berczy Park                                                            58   
Birch Cliff, Cliffside West                                             4   
Brockton, Parkdale Village, Exhibition Place                           23   
Business reply mail Processing Centre, South Ce...                     17   
CN Tower, King and Spadina, Railway Lands, Harb...                     14   
Caledonia-Fairbanks                                                     4   
Canada Post Gateway Processing Centre                                  12   
Cedarbrae                                                               8   
Central Bay Street                                                     62   
Christie                                                               17   
Church and Wellesley                                                   77   
Clarks Corners, Tam O'Shanter, Sullivan                                13   
Cliffside, Cliffcrest, Scarborough Village West                         3   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             33   
Davisville North                                                        7   
Del Ray, Mount Dennis, Keelsdale and Silverthorn                        4   
Don Mills                                                              27   
Dorset Park, Wexford Heights, Scarborough Town ...                      6   
Downsview                                                              18   
Dufferin, Dovercourt Village                                           16   
East Toronto, Broadview North (Old East York)                           3   
Eringate, Bloordale Gardens, Old Burnhamthorpe,...                      7   
Fairview, Henry Farm, Oriole                                           64   
First Canadian Place, Underground city                                100   
...                                                                   ...   
Rouge Hill, Port Union, Highland Creek                                  2   
Runnymede, Swansea                                                     42   
Runnymede, The Junction North                                           5   
Scarborough Village                                                     2   
South Steeles, Silverstone, Humbergate, Jamesto...                      8   
St. James Town                                                         80   
St. James Town, Cabbagetown                                            42   
Steeles West, L'Amoreaux West                                          15   
Stn A PO Boxes                                                         93   
Studio District                                                        41   
Summerhill West, Rathnelly, South Hill, Forest ...                     17   
The Annex, North Midtown, Yorkville                                    22   
The Beaches                                                             6   
The Danforth West, Riverdale                                           42   
The Kingsway, Montgomery Road, Old Mill North                           2   
Thorncliffe Park                                                       19   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         34   
Victoria Village                                                        5   
West Deane Park, Princess Gardens, Martin Grove.

#### How many categorys can we find?

In [208]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighbourhood'] = toronto_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighbourhood  Accessories Store  \
0                          Malvern, Rouge                  0   
1  Rouge Hill, Port Union, Highland Creek                  0   
2  Rouge Hill, Port Union, Highland Creek                  0   
3       Guildwood, Morningside, West Hill                  0   
4       Guildwood, Morningside, West Hill                  0   

   Afghan Restaurant  Airport  Airport Food Court  Airport Lounge  \
0                  0        0                   0               0   
1                  0        0                   0               0   
2                  0        0                   0               0   
3                  0        0                   0               0   
4                  0        0                   0               0   

   Airport Service  Airport Terminal  American Restaurant  Antique Shop  ...  \
0                0                 0                    0             0  ...   
1                0                 0                    0             0  ...   
2                0                 0                    0             0  ...   
3                0                 0                    0             0  ...   
4                0                 0                    0             0  ...   

   Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0              0                              0                 0   
1              0                              0                 0   
2              0                              0                 0   
3              0                              0                 0   
4              0                              0                 0   

   Video Store  Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0            0                      0                0         0            0   
1            0                      0                0         0            0   
2            0                      0                0         0            0   
3            0                      0                0         0            0   
4            0                      0                0         0            0   

   Women's Store  Yoga Studio  
0              0            0  
1              0            0  
2              0            0  
3              0            0  
4              0            0  

[5 rows x 272 columns]

In [209]:
toronto_onehot.shape

(2136, 272)

In [210]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Afghan Restaurant  Airport  Airport Food Court  Airport Lounge  \
0                0.0      0.0                 0.0             0.0   
1                0.0      0.0                 0.0             0.0   
2                0.0      0.0                 0.0             0.0   
3                0.0      0.0                 0.0             0.0   
4                0.0      0.0                 0.0             0.0   

   Airport Service  Airport Terminal  American Restaurant  Antique Shop  ...  \
0              0.0               0.0             0.000000           0.0  ...   
1              0.0               0.0             0.000000           0.0  ...   
2              0.0               0.0             0.000000           0.0  ...   
3              0.0               0.0             0.000000           0.0  ...   
4              0.0               0.0             0.043478           0.0  ...   

   Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0            0.0                            0.0               0.0   
1            0.0                            0.0               0.0   
2            0.0                            0.0               0.0   
3            0.0                            0.0               0.0   
4            0.0                            0.0               0.0   

   Video Store  Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0     0.000000                    0.0              0.0       0.0          0.0   
1     0.000000                    0.0              0.0       0.0          0.0   
2     0.047619                    0.0              0.0       0.0          0.0   
3     0.000000                    0.0              0.0       0.0          0.0   
4     0.000000                    0.0              0.0       0.0          0.0   

   Women's Store  Yoga Studio  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  

[5 rows x 272 columns]

In [211]:
toronto_grouped.shape

(97, 272)

#### Get the top 10 for each neighbourhood

In [213]:

def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [214]:
import numpy as np
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                  Bank   
3                                  Bayview Village                  Café   
4                Bedford Park, Lawrence Manor East           Coffee Shop   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0  Latin American Restaurant        Breakfast Spot          Skating Rink   
1               Dance Studio                   Gym           Coffee Shop   
2                Coffee Shop         Shopping Mall              Pharmacy   
3                       Bank    Chinese Restaurant   Japanese Restaurant   
4                 Restaurant      Sushi Restaurant    Italian Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0      Doner Restaurant    Dim Sum Restaurant                 Diner   
1          Skating Rink              Pharmacy                   Pub   
2           Supermarket      Sushi Restaurant            Restaurant   
3      Department Store    Dim Sum Restaurant                 Diner   
4        Sandwich Place              Pharmacy                   Pub   

  8th Most Common Venue 9th Most Common Venue     10th Most Common Venue  
0        Discount Store   Distribution Center                    Dog Run  
1        Sandwich Place    Athletics & Sports                       Pool  
2         Deli / Bodega   Fried Chicken Joint  Middle Eastern Restaurant  
3        Discount Store   Distribution Center                    Dog Run  
4          Liquor Store               Butcher            Thai Restaurant

In [275]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1], dtype=int32)

#### Merge the dataframe with the top 10 and the cluster for each neighbourhood

In [277]:
# add clustering labels
#neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = df_toronto_merge

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Postal Code      Borough                            Neighborhood   Latitude  \
0         M1B  Scarborough                          Malvern, Rouge  43.806686   
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2         M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3         M1G  Scarborough                                  Woburn  43.770992   
4         M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.194353             2.0  Fast Food Restaurant         Deli / Bodega   
1 -79.160497             1.0                   Bar        History Museum   
2 -79.188711             1.0    Mexican Restaurant     Electronics Store   
3 -79.216917             1.0           Coffee Shop     Korean Restaurant   
4 -79.239476             1.0      Hakka Restaurant           Gas Station   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0           Event Space  Ethiopian Restaurant     Electronics Store   
1           Yoga Studio      Doner Restaurant    Dim Sum Restaurant   
2        Breakfast Spot                  Bank   Rental Car Location   
3           Yoga Studio      Doner Restaurant    Dim Sum Restaurant   
4                  Bank   Fried Chicken Joint    Athletics & Sports   

         6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0  Eastern European Restaurant             Drugstore            Donut Shop   
1                        Diner        Discount Store   Distribution Center   
2               Medical Center         Moving Target          Intersection   
3                        Diner        Discount Store   Distribution Center   
4              Thai Restaurant  Caribbean Restaurant                Bakery   

  9th Most Common Venue 10th Most Common Venue  
0      Doner Restaurant                Dog Run  
1               Dog Run             Donut Shop  
2           Yoga Studio                  Diner  
3               Dog Run              Drugstore  
4      Doner Restaurant                Dog Run

In [249]:
toronto_merged[toronto_merged['Cluster Labels'].isnull()]

Postal Code      Borough                             Neighborhood  \
16         M1X  Scarborough                              Upper Rouge   
93         M9A    Etobicoke  Islington Avenue, Humber Valley Village   

    Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
16             NaN                   NaN                   NaN   
93             NaN                   NaN                   NaN   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
16                   NaN                   NaN                   NaN   
93                   NaN                   NaN                   NaN   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
16                   NaN                   NaN                   NaN   
93                   NaN                   NaN                   NaN   

   9th Most Common Venue 10th Most Common Venue  
16                   NaN                    NaN  
93                   NaN                    NaN

####  Plot the clusters in the map

In [278]:
import matplotlib.cm as cm
import matplotlib.colors as colors

# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighborhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

#### Examine Cluster

##### Cluster 1

In [280]:

toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
14       Scarborough             0.0                  Park   
23        North York             0.0                  Park   
25        North York             0.0                  Park   
40         East York             0.0                  Park   
44   Central Toronto             0.0                  Park   
50  Downtown Toronto             0.0                  Park   
74              York             0.0                  Park   
79        North York             0.0                  Park   
90         Etobicoke             0.0                  Park   
98              York             0.0                  Park   

   2nd Most Common Venue       3rd Most Common Venue 4th Most Common Venue  \
14            Playground                 Yoga Studio               Dog Run   
23     Convenience Store                 Yoga Studio      Doner Restaurant   
25     Food & Drink Shop                 Yoga Studio               Dog Run   
40           Coffee Shop           Convenience Store           Yoga Studio   
44           Swim School                    Bus Line           Yoga Studio   
50                 Trail                  Playground      Department Store   
74                  Pool               Women's Store               Dog Run   
79                Bakery  Construction & Landscaping           Yoga Studio   
90                 River                 Yoga Studio   Distribution Center   
98     Convenience Store                 Yoga Studio      Doner Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
14      Department Store          Dessert Shop    Dim Sum Restaurant   
23          Dessert Shop    Dim Sum Restaurant                 Diner   
25          Dessert Shop    Dim Sum Restaurant                 Diner   
40      Doner Restaurant    Dim Sum Restaurant                 Diner   
44   Distribution Center          Dessert Shop    Dim Sum Restaurant   
50          Dessert Shop    Dim Sum Restaurant                 Diner   
74      Department Store          Dessert Shop    Dim Sum Restaurant   
79      Doner Restaurant    Dim Sum Restaurant                 Diner   
90      Department Store          Dessert Shop    Dim Sum Restaurant   
98          Dessert Shop    Dim Sum Restaurant                 Diner   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
14                 Diner        Discount Store    Distribution Center  
23        Discount Store   Distribution Center                Dog Run  
25        Discount Store   Distribution Center       Doner Restaurant  
40        Discount Store   Distribution Center                Dog Run  
44                 Diner        Discount Store                Dog Run  
50        Discount Store   Distribution Center                Dog Run  
74                 Diner        Discount Store    Distribution Center  
79        Discount Store   Distribution Center                Dog Run  
90                 Diner        Discount Store                Dog Run  
98        Discount Store   Distribution Center                Dog Run

##### Cluster 2

In [281]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]


Borough  Cluster Labels      1st Most Common Venue  \
1         Scarborough             1.0                        Bar   
2         Scarborough             1.0         Mexican Restaurant   
3         Scarborough             1.0                Coffee Shop   
4         Scarborough             1.0           Hakka Restaurant   
6         Scarborough             1.0             Discount Store   
7         Scarborough             1.0                     Bakery   
8         Scarborough             1.0               Skating Rink   
9         Scarborough             1.0            College Stadium   
10        Scarborough             1.0          Indian Restaurant   
11        Scarborough             1.0  Middle Eastern Restaurant   
12        Scarborough             1.0                     Lounge   
13        Scarborough             1.0                Pizza Place   
15        Scarborough             1.0         Chinese Restaurant   
17         North York             1.0                       Pool   
18         North York             1.0             Clothing Store   
19         North York             1.0                       Café   
20         North York             1.0                  Cafeteria   
21         North York             1.0                  Piano Bar   
22         North York             1.0           Ramen Restaurant   
24         North York             1.0                   Pharmacy   
26         North York             1.0        Japanese Restaurant   
27         North York             1.0        Japanese Restaurant   
28         North York             1.0                       Bank   
29         North York             1.0              Metro Station   
30         North York             1.0              Grocery Store   
31         North York             1.0              Grocery Store   
32         North York             1.0              Grocery Store   
33         North York             1.0              Grocery Store   
34         North York             1.0      Portuguese Restaurant   
35          East York             1.0                Pizza Place   
..                ...             ...                        ...   
66   Downtown Toronto             1.0                       Café   
67   Downtown Toronto             1.0                       Café   
68   Downtown Toronto             1.0             Airport Lounge   
69   Downtown Toronto             1.0                Coffee Shop   
70   Downtown Toronto             1.0                Coffee Shop   
71         North York             1.0             Clothing Store   
72         North York             1.0                     Bakery   
73               York             1.0               Hockey Arena   
75   Downtown Toronto             1.0              Grocery Store   
76       West Toronto             1.0                   Pharmacy   
77       West Toronto             1.0                        Bar   
78       West Toronto             1.0                       Café   
80               York             1.0             Sandwich Place   
81               York             1.0             Breakfast Spot   
82       West Toronto             1.0         Mexican Restaurant   
83       West Toronto             1.0             Breakfast Spot   
84       West Toronto             1.0                Coffee Shop   
85   Downtown Toronto             1.0                Coffee Shop   
86        Mississauga             1.0                      Hotel   
87       East Toronto             1.0         Light Rail Station   
88          Etobicoke             1.0                Coffee Shop   
89          Etobicoke             1.0                Pizza Place   
92          Etobicoke             1.0              Grocery Store   
94          Etobicoke             1.0                  Gift Shop   
95          Etobicoke             1.0                   Pharmacy   
96         North York             1.0              Shopping Mall   
99          Etobicoke             1.0                Pizza Place   
10

##### cluster 3

In [282]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0  Scarborough             2.0  Fast Food Restaurant         Deli / Bodega   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0           Event Space  Ethiopian Restaurant     Electronics Store   

         6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0  Eastern European Restaurant             Drugstore            Donut Shop   

  9th Most Common Venue 10th Most Common Venue  
0      Doner Restaurant                Dog Run

##### cluster 4

In [284]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
5  Scarborough             3.0            Playground     Convenience Store   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
5           Yoga Studio      Doner Restaurant          Dessert Shop   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
5    Dim Sum Restaurant                 Diner        Discount Store   

  9th Most Common Venue 10th Most Common Venue  
5   Distribution Center                Dog Run

In [285]:
##### cluster 5

In [286]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels          1st Most Common Venue  \
91   Etobicoke             4.0                 Baseball Field   
97  North York             4.0  Paper / Office Supplies Store   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
91           Yoga Studio               Dog Run          Dessert Shop   
97        Baseball Field           Yoga Studio      Doner Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
91    Dim Sum Restaurant                 Diner        Discount Store   
97    Dim Sum Restaurant                 Diner        Discount Store   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
91   Distribution Center      Doner Restaurant         Farmers Market  
97   Distribution Center               Dog Run             Donut Shop